File: backend/app/api/ingredients.py
"""Ingredient API router — client and admin endpoints."""

from uuid import UUID

from fastapi import APIRouter, Depends, Query

from app.core.dependencies import (
    PaginationParams,
    get_current_admin,
    get_current_user,
    get_ingredient_service,
    get_pagination,
)
from app.models.ingredient import Ingredient
from app.models.user import User
from app.schemas.ingredient import (
    IngredientAdminResponse,
    IngredientCreate,
    IngredientDeleteResponse,
    IngredientResponse,
    IngredientUpdate,
)
from app.schemas.pagination import PaginatedResponse
from app.services.ingredient import IngredientService

router = APIRouter(prefix="/ingredients", tags=["ingredients"])


@router.get("", response_model=PaginatedResponse[IngredientResponse], status_code=200)
async def list_ingredients(
    pagination: PaginationParams = Depends(get_pagination),
    search: str | None = Query(None),
    slug: str | None = Query(None),
    _user: User = Depends(get_current_user),
    service: IngredientService = Depends(get_ingredient_service),
) -> PaginatedResponse[Ingredient]:
    return await service.list(pagination, search=search, slug=slug, is_active=True)


@router.get("/admin", response_model=PaginatedResponse[IngredientAdminResponse], status_code=200)
async def list_ingredients_admin(
    pagination: PaginationParams = Depends(get_pagination),
    search: str | None = Query(None),
    slug: str | None = Query(None),
    is_active: bool | None = Query(None),
    _admin: User = Depends(get_current_admin),
    service: IngredientService = Depends(get_ingredient_service),
) -> PaginatedResponse[Ingredient]:
    return await service.list(pagination, search=search, slug=slug, is_active=is_active)


@router.get("/{ingredient_id}/admin", response_model=IngredientAdminResponse, status_code=200)
async def get_ingredient_admin(
    ingredient_id: UUID,
    _admin: User = Depends(get_current_admin),
    service: IngredientService = Depends(get_ingredient_service),
) -> Ingredient:
    return await service.get_by_id(ingredient_id)


@router.post("/admin", response_model=IngredientAdminResponse, status_code=201)
async def create_ingredient_admin(
    data: IngredientCreate,
    _admin: User = Depends(get_current_admin),
    service: IngredientService = Depends(get_ingredient_service),
) -> Ingredient:
    return await service.create(data)


@router.patch("/{ingredient_id}/admin", response_model=IngredientAdminResponse, status_code=200)
async def update_ingredient_admin(
    ingredient_id: UUID,
    data: IngredientUpdate,
    _admin: User = Depends(get_current_admin),
    service: IngredientService = Depends(get_ingredient_service),
) -> Ingredient:
    return await service.update(ingredient_id, data)


@router.delete("/{ingredient_id}/admin", response_model=IngredientDeleteResponse, status_code=200)
async def delete_ingredient_admin(
    ingredient_id: UUID,
    _admin: User = Depends(get_current_admin),
    service: IngredientService = Depends(get_ingredient_service),
) -> IngredientDeleteResponse:
    await service.delete(ingredient_id)
    return IngredientDeleteResponse(id=ingredient_id, is_deleted=True)


@router.get("/{ingredient_id}", response_model=IngredientResponse, status_code=200)
async def get_ingredient(
    ingredient_id: UUID,
    _user: User = Depends(get_current_user),
    service: IngredientService = Depends(get_ingredient_service),
) -> Ingredient:
    return await service.get_by_id(ingredient_id)

File: frontend/src/api/ingredients.ts
import apiClient from "./client";
import type {
  IngredientAdmin,
  IngredientCreate,
  IngredientDeleteResponse,
  IngredientUpdate,
  PaginatedResponse,
} from "@/types";

export async function listIngredients(
  limit = 20,
  offset = 0,
  search?: string,
): Promise<PaginatedResponse<IngredientAdmin>> {
  const params: Record<string, unknown> = { limit, offset };
  if (search) params.search = search;
  const res = await apiClient.get<PaginatedResponse<IngredientAdmin>>("/ingredients/admin", { params });
  return res.data;
}

export async function getIngredient(id: string): Promise<IngredientAdmin> {
  const res = await apiClient.get<IngredientAdmin>(`/ingredients/${id}/admin`);
  return res.data;
}

export async function createIngredient(data: IngredientCreate): Promise<IngredientAdmin> {
  const res = await apiClient.post<IngredientAdmin>("/ingredients/admin", data);
  return res.data;
}

export async function updateIngredient(id: string, data: IngredientUpdate): Promise<IngredientAdmin> {
  const res = await apiClient.patch<IngredientAdmin>(`/ingredients/${id}/admin`, data);
  return res.data;
}

export async function deleteIngredient(id: string): Promise<IngredientDeleteResponse> {
  const res = await apiClient.delete<IngredientDeleteResponse>(`/ingredients/${id}/admin`);
  return res.data;
}
